internal/pkg/gateway/api.go

```diff
@@ -273,6 +270,44 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
 	return endorseResponse, nil
 }
 
+// responseStatus unpacks the proposal response and error values that are returned from ProcessProposal and
+// determines how the gateway should react (retry?, close connection?).
+// Uses the grpc canonical status error codes and their recommended actions.
+// Returns:
+// - response successful (bool)
+// - error message extracted from the err or generated from 500 proposal response
+// - should the gateway retry (only the Evaluate() uses this)
+// - should the gateway close the connection and remove the peer from its registry
+func (gs *Server) responseStatus(response *peer.ProposalResponse, err error) (bool, string, bool, bool) {
+	// if the ProcessProposal function returned an err, then it must have been from the unpackProposal() or preProcess() stages
```

Comment:
grpc issues will also return an err, right?

Reply:
Yes, I think I handle that in the err!=nil && response==nil case